akka-docs/src/main/paradox/typed/persistence.md

```diff
@@ -216,6 +212,73 @@ Scala
 
+## Effects and Side Effects
+
+A command handler returns an `Effect` directive that defines what event or events, if any, to persist. 
+Effects are created using @java[a factory that is returned via the `Effect()` method] @scala[the `Effect` factory]
+and can be one of: 
+
+* `persist` will persist one single event or several events atomically, i.e. all events
+  are stored or none of them are stored if there is an error
+* `none` no events are to be persisted, for example a read-only command
+* `unhandled` the command is unhandled (not supported) in current state
+* `stop` stop this actor
+* `stash` the current command is stashed
+* `unstashAll` process the commands that were stashed with @scala[`Effect.stash`]@java[`Effect().stash`]
+* `reply` send a reply message to the given `ActorRef`
+
+Note that there is only one of these. It is not possible to both persist and say none/unhandled.
+
+In addition to returning the primary `Effect` for the command `EventSourcedBehavior`s can also 
+chain side effects that are to be performed after successful persist which is achieved with the `thenRun`
+function e.g @scala[`Effect.persist(..).thenRun`]@java[`Effect().persist(..).thenRun`].
+
+In the example below the state is sent to the `subscriber` ActorRef. Note that the new state after applying 
+the event is passed as parameter of the `thenRun` function.
+
+All `thenRun` registered callbacks are executed sequentially after successful execution of the persist statement
+(or immediately, in case of `none` and `unhandled`).
+
+In addition to `thenRun` the following actions can also be performed after successful persist:
+
+* `thenStop` the actor will be stopped
+* `thenUnstashAll` process the commands that were stashed with @scala[`Effect.stash`]@java[`Effect().stash`]
+* `thenReply` send a reply message to the given `ActorRef`
+
+Example of effects:
+
+Scala
+:  @@snip [BasicPersistentBehaviorCompileOnly.scala](/akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/BasicPersistentBehaviorCompileOnly.scala) { #effects }
+
+Java
+:  @@snip [BasicPersistentBehaviorTest.java](/akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/BasicPersistentBehaviorTest.java) { #effects }
+
+Most of the time this will be done with the `thenRun` method on the `Effect` above. You can factor out
+common side effects into functions and reuse for several commands. For example:
+
+Scala
+:  @@snip [PersistentActorCompileOnlyTest.scala](/akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentActorCompileOnlyTest.scala) { #commonChainedEffects }
+
+Java
+:  @@snip [PersistentActorCompileOnlyTest.java](/akka-persistence-typed/src/test/java/akka/persistence/typed/javadsl/PersistentActorCompileOnlyTest.java) { #commonChainedEffects }
+
+### Side effects ordering and guarantees
+
+Any side effects are executed on an at-once basis and will not be executed if the persist fails.
```

Comment:
Maybe point out that they are not executed on replay but only on original persist, not that it is unclear but can't hurt?

Reply:
added some more here, worth reviewing again